syntax = "proto3";

package aea.eightballer.ohlcv.v0_1_0;

message OhlcvMessage{

  // Custom Types
  message ErrorCode{
    enum ErrorCodeEnum {
      ERROR_CODE_ENUM_UNSUPPORTED_PROTOCOL = 0;
      ERROR_CODE_ENUM_DECODING_ERROR = 1;
      ERROR_CODE_ENUM_INVALID_MESSAGE = 2;
      ERROR_CODE_ENUM_UNSUPPORTED_SKILL = 3;
      ERROR_CODE_ENUM_INVALID_DIALOGUE = 4;
    }
    ErrorCodeEnum error_code = 1;
  }


  // Performatives and contents
  message Subscribe_Performative{
    string exchange_id = 1;
    string market_name = 2;
    int32 interval = 3;
  }

  message Candlestick_Performative{
    string exchange_id = 1;
    string market_name = 2;
    int32 interval = 3;
    double open = 4;
    double high = 5;
    double low = 6;
    double close = 7;
    double volume = 8;
    int32 timestamp = 9;
  }

  message History_Performative{
    string exchange_id = 1;
    string market_name = 2;
    int32 start_timestamp = 3;
    int32 end_timestamp = 4;
    int32 interval = 5;
  }

  message Error_Performative{
    ErrorCode error_code = 1;
    string error_msg = 2;
    map<string, bytes> error_data = 3;
  }

  message End_Performative{
  }


  oneof performative{
    Candlestick_Performative candlestick = 5;
    End_Performative end = 6;
    Error_Performative error = 7;
    History_Performative history = 8;
    Subscribe_Performative subscribe = 9;
  }
}
